### ISSUE TITLE
Unable to run on Monterey on M1 MacBook Pro

### ISSUE THREAD

**Issue by wintermute-motherbrain** [ApsarasX/llvm-bindings#9]
- OS version: Mac OS 12.0.1 M1 MacBook Pro 
- Node.js version: 17.01
- LLVM version: 13

Was able to successfully install the npm module on a brand new M1 MacBook but run into runtime errors when trying to launch:

`Error: dlopen(node_modules/llvm-bindings/build/Release/llvm-bindings.node, 0x0001): symbol not found in flat namespace '_LLVMInitializeAArch64AsmParser`

**Expected behavior**

Everything works fine on an Intel Mac.  I am assuming that there are some issues with M1 Mac's being Arm?

**Comment by wintermute-motherbrain** [comment]
I am wondering if recompiling will actually solve this, likely the release libs of llvm-bindings are compiled for different architecture?  I'll try recompling the libs from scratch and see if that fixes it.

**Comment by wintermute-motherbrain** [comment]
Ah, I see that its only been tested with x64.  Will close this issue for now.

**Comment by wintermute-motherbrain** [comment]
Dug deeper, thought it was the bindings and llvm being built for the wrong arch but built llvm-bindings and llvm from scratch and ensured arch was aarch64 and the issue still occurs.  Interesting enough llvm-node does work on M1...not sure if there are any plans to support the M1.

**Comment by ApsarasX** [comment]
Thanks for your feedback. Although I have not tested `llvm-bindings` with M1 MacBook Pro, this error is indeed unexpected. I will fix this error as soon as possible.

**Comment by ApsarasX** [comment]
I test `llvm-bindings` on Ubuntu 18.04(arm). But I didn't reproduce the problem you got. If you give some sample code, I will better locate the problem

**Comment by wintermute-motherbrain** [comment]
Hi,

This is what I did to reproduce...

On fresh M1 Mac:
```
brew install llvm
brew install cmake
brew install nodejs
brew install npm
git clone https://github.com/ApsarasX/llvm-bindings.git
cd llvm-bindings
npm config set cmake_LLVM_DIR /opt/homebrew/opt/llvm/lib/cmake/llvm
npm install
```
After running the above, everything compiles and builds fun.  Then execute the following and you'll get the dopen issue.

```
jpaiva@Cyberspace7 llvm-bindings % npm test

> llvm-bindings@0.3.1 pretest
> npm run build:debug


> llvm-bindings@0.3.1 build:debug
> cross-env CMAKE_BUILD_PARALLEL_LEVEL=$(node -p 'require(`os`).cpus().length') cmake-js build -D

[
  '/Users/jpaiva/.nvm/versions/node/v17.2.0/bin/node',
  '/Users/jpaiva/dev/llvm-bindings/node_modules/.bin/cmake-js',
  'build',
  '-D'
]
info TOOL Using Unix Makefiles generator.
info CMD BUILD
info RUN [
info RUN   'cmake',
info RUN   '--build',
info RUN   '/Users/jpaiva/dev/llvm-bindings/build',
info RUN   '--config',
info RUN   'Debug'
info RUN ]
[100%] Built target llvm-bindings

> llvm-bindings@0.3.1 test
> ts-node test/index.ts

Error: dlopen(/Users/jpaiva/dev/llvm-bindings/build/Release/llvm-bindings.node, 0x0001): symbol not found in flat namespace '_LLVMInitializeAArch64AsmParser'
    at Object.Module._extensions..node (node:internal/modules/cjs/loader:1179:18)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at bindings (/Users/jpaiva/dev/llvm-bindings/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/Users/jpaiva/dev/llvm-bindings/index.js:1:37)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Module.load (node:internal/modules/cjs/loader:975:32) {
  code: 'ERR_DLOPEN_FAILED'
}
```

I've tried brew nodejs install as well as using nvm with a couple different versions and all failed with the exact same error.  I've also built even the latest llvm (14) and got it to compile with llvm-bindings but still same dlopen issue.

Regarding Ubuntu - I tried compiling on an Ubuntu machine that is Arm64 (I have a raspberry pi 4) and I ran into napi compile issues when compiling llvm-bindings so I was never able to test Arm64 ubuntu.  I got the same napi problems on the M1 with Ubuntu VM as well as on a the Raspberry Pi 4.  These seem different from the dlopen issue with M1.

What version of Nodejs are you using?

**Comment by wintermute-motherbrain** [comment]
I managed to get a hold of another M1 and was able to install and run so it seems its related to the environment.  The M1 that worked was using node v16 and npm v7.  They were also on Big Sur and not Monterey - its quite likely its related to MacOS 12.0.1 (Monterey).

**Comment by ApsarasX** [comment]
I have tested the following environments without problems:
- Ubuntu 20.04 ARM64, Node.js 16
- Ubuntu 20.04 ARM64, Node.js 17
- Ubuntu 18.04 ARM64, Node.js 16
- Ubuntu 18.04 ARM64, Node.js 17

But I don't have a M1 MacBook, so I can't solve this problem for the time being.

**Comment by ApsarasX** [comment]
If you got an error about `napi_object_freeze` on Ubuntu arm64, please refer to #6

**Comment by ApsarasX** [comment]
This issue may be related to Monterey, I also found a similar error on stackoverflow

https://stackoverflow.com/questions/69947571/updating-to-ruby-3-0-x-with-rails-symbol-not-found-in-flat-namespace-rhash-em

**Comment by wintermute-motherbrain** [comment]
Yep, looks like the issue is with Monterey and unfortunately no clear path forward for the time being.

In the meantime, I was able to get Ubuntu Arm64 working after fixing NodeJS.  I am now using Ubuntu 20.04 ARM64 via Multipass on the M1 with no problems.  Guess I'll work in this environment till there is a solution for Monterey.

Thanks for the help!

**Comment by ApsarasX** [comment]
I also encountered the problem you encountered.

In the end I found that this was because I did not add the LLVM related link component  in `LLVM.cmake`.

I have fixed this problem in the `v0.3.3` release, you can try again.

At the same time, I blocked the function:
- `InitializeAllTargetInfos`
- `InitializeAllTargets`
- `InitializeAllTargetMCs`
- `InitializeAllAsmPrinters`
- `InitializeAllAsmParsers`
- `InitializeAllDisassemblers`

If you want to initialize target info on your platform, please consider calling the following function:
- `InitializeNativeTarget`
- `InitializeNativeTargetAsmPrinter`
- `InitializeNativeTargetAsmParser`
- `InitializeNativeTargetDisassembler`

**Comment by wintermute-motherbrain** [comment]
Nice!  Confirmed that version 0.3.3 is working on Monterey with M1 Mac.

**Comment by ApsarasX** [comment]
These previously blocked functions have been unblocked in the commit 1bbf8cf40d7d7e3ce82bbc1c864a25a6d4ebb9a2, and will be released in the next version(v0.3.4).
- `InitializeAllTargetInfos`
- `InitializeAllTargets`
- `InitializeAllTargetMCs`
- `InitializeAllAsmPrinters`
- `InitializeAllAsmParsers`
- `InitializeAllDisassemblers`
